### PR TITLE
impl(oauth2): helper functions to compute `HMAC_SHA256()`

### PR DESCRIPTION
**Pull request by coryan**
As part of the AWS external account support we need to generate a signed request:

> Construct the AWS signed request (AWS Signature Version 4) using the
> `GetCallerIdentity`

(from)
https://google.aip.dev/auth/4117#determining-the-subject-token-in-aws

These signed requests are documented in:

https://docs.aws.amazon.com/general/latest/gr/create-signed-request.html

They require computing multiple `HMAC_SHA256()` values. The implementation is simple if you already have a dependency on OpenSSL (as we do). A longer description of these can be found in Wikipedia:

https://en.wikipedia.org/wiki/HMAC

Part of the work for #5915

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10442)

### CODE REVIEW

File: google/cloud/internal/sha256_hmac.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_SHA256_HMAC_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_SHA256_HMAC_H
+
+#include "google/cloud/internal/sha256_type.h"
+#include "google/cloud/version.h"
+#include "absl/types/span.h"
+#include <array>
```
Comment (devbww) [inline google/cloud/internal/sha256_hmac.h:21]: Unused.

File: google/cloud/internal/sha256_hmac.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#include <cstdint>
+#include <string>
+#include <vector>
```
Comment (devbww) [inline google/cloud/internal/sha256_hmac.h:24]: Unused.

File: google/cloud/internal/sha256_hmac_test.cc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  auto data = std::string{"The quick brown fox jumps over the lazy dog"};
+  // Compare against the HEX-encoded value just because they are easier to read.
+  auto const expected = std::string{
+      "f7bc83f430538424b13298e6aa6fb143ef4d59a14946175997479dbc2d1a3cd8"};
+  EXPECT_EQ(expected, HexEncode(Sha256Hmac("key", data)));
+  EXPECT_EQ(expected, HexEncode(Sha256Hmac("key", absl::Span<char>(data))));
+  EXPECT_EQ(expected,
+            HexEncode(Sha256Hmac("key", absl::Span<char const>(data))));
+  std::vector<std::uint8_t> v(data.begin(), data.end());
+  EXPECT_EQ(expected,
```
Comment (devbww) [inline google/cloud/internal/sha256_hmac_test.cc:39]: `EXPECT_EQ(expected, HexEncode(Sha256Hmac("key", v)));` too, like for `data`?

File: google/cloud/internal/sha256_hmac_test.cc [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  auto data = std::string{"The quick brown fox jumps over the lazy dog"};
+  // Compare against the HEX-encoded value just because they are easier to read.
+  auto const expected = std::string{
+      "f61f354e69e4e35dd936913f098993f4a07254ef9c156e26842f07e998d6e61e"};
+  auto key = Sha256Hmac("key", data);
+  EXPECT_EQ(expected, HexEncode(Sha256Hmac(key, data)));
+  EXPECT_EQ(expected, HexEncode(Sha256Hmac(key, absl::Span<char>(data))));
+  EXPECT_EQ(expected, HexEncode(Sha256Hmac(key, absl::Span<char const>(data))));
+  std::vector<std::uint8_t> v(data.begin(), data.end());
+  EXPECT_EQ(expected, HexEncode(Sha256Hmac(key, absl::Span<std::uint8_t>(v))));
```
Comment (devbww) [inline google/cloud/internal/sha256_hmac_test.cc:60]: Ditto.

File: google/cloud/internal/sha256_hmac.cc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+Sha256Type Sha256Hmac(std::string const& key, absl::Span<char const> data) {
+  return Sha256HmacImpl(absl::Span<char const>(key),
+                        reinterpret_cast<unsigned char const*>(data.data()),
+                        data.size());
+}
+
+Sha256Type Sha256Hmac(std::string const& key,
+                      absl::Span<std::uint8_t const> data) {
+  return Sha256HmacImpl(absl::Span<char const>(key),
+                        reinterpret_cast<unsigned char const*>(data.data()),
```
Comment (devbww) [inline google/cloud/internal/sha256_hmac.cc:54]: It is worth noting (in our heads, if nowhere else) that if `std::uint8_t` exists, as we assume, that it must be a synonym for `unsigned char`.  So, the cast is fine.

File: google/cloud/internal/sha256_hmac.cc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+namespace internal {
+namespace {
+
+static_assert(EVP_MAX_MD_SIZE >= Sha256Type().size(),
+              "EVP_MAX_MD_SIZE is too small");
+
+template <typename T>
+Sha256Type Sha256HmacImpl(absl::Span<T const> key, unsigned char const* data,
+                          std::size_t count) {
+  std::array<unsigned char, EVP_MAX_MD_SIZE> digest;
```
Comment (devbww) [inline google/cloud/internal/sha256_hmac.cc:32]: `#include <array>`

File: google/cloud/internal/sha256_hmac.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/sha256_hmac.h]: Removed

File: google/cloud/internal/sha256_hmac.h (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/sha256_hmac.h]: Removed

File: google/cloud/internal/sha256_hmac_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/sha256_hmac_test.cc:40]: Done

File: google/cloud/internal/sha256_hmac_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/internal/sha256_hmac_test.cc:62]: Done

File: google/cloud/internal/sha256_hmac.cc (hunk 5 above)
Comment (devbww) [reply google/cloud/internal/sha256_hmac.cc:58]: "When `std::uint8_t` exists it must be `unsigned char`" perhaps?

**Review by devbww — APPROVED** [review]

**Pull request merged**